src/pages/workspace/WorkspaceNewRoomPage.js

```diff
@@ -0,0 +1,182 @@
+import _ from 'underscore';
+import {withOnyx} from 'react-native-onyx';
+import PropTypes from 'prop-types';
+import withFullPolicy, {fullPolicyDefaultProps, fullPolicyPropTypes} from './withFullPolicy';
+import withLocalize, {withLocalizePropTypes} from '../../components/withLocalize';
+import compose from '../../libs/compose';
+import HeaderWithCloseButton from '../../components/HeaderWithCloseButton';
+import Navigation from '../../libs/Navigation/Navigation';
+import ScreenWrapper from '../../components/ScreenWrapper';
+import styles from '../../styles/styles';
+import TextInputWithLabel from '../../components/TextInputWithLabel';
+import ExpensiPicker from '../../components/ExpensiPicker';
+import ONYXKEYS from '../../ONYXKEYS';
+import CONST from '../../CONST';
+import Button from '../../components/Button';
+import FixedFooter from '../../components/FixedFooter';
+import {createPolicyRoom} from '../../libs/actions/Report';
+
+const propTypes = {
+    /** All reports shared with the user */
+    reports: PropTypes.shape({
+        reportName: PropTypes.string,
+        type: PropTypes.string,
+        policyID: PropTypes.string,
+    }).isRequired,
+
+    ...fullPolicyPropTypes,
+
+    ...withLocalizePropTypes,
+};
+const defaultProps = {
+    betas: [],
+    ...fullPolicyDefaultProps,
+};
+
+class WorkspaceNewRoomPage extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            roomName: '',
+            policyID: '',
+            visibility: CONST.REPORT.VISIBILITY.RESTRICTED,
+            isLoading: false,
+            error: '',
+            workspaceOptions: [],
+        };
+        this.onWorkspaceSelect = this.onWorkspaceSelect.bind(this);
+        this.onSubmit = this.onSubmit.bind(this);
+        this.checkAndModifyRoomName = this.checkAndModifyRoomName.bind(this);
+    }
+
+    componentDidUpdate(prevProps) {
+        if (this.props.policies.length === prevProps.policies.length) {
+            return;
+        }
+        const workspaces = _.filter(this.props.policies, policy => policy && policy.type === CONST.POLICY.TYPE.FREE);
+
+        // eslint-disable-next-line react/no-did-update-set-state
+        this.setState({workspaceOptions: _.map(workspaces, policy => ({label: policy.name, key: policy.id, value: policy.id}))});
+    }
+
+    /**
+     * Called when a workspace is selected. Also calls checkAndModifyRoomName,
+     * which displays an error if the given roomName exists on the newly selected workspace.
+     * @param {String} policyID
+     */
+    onWorkspaceSelect(policyID) {
+        this.setState({policyID});
+        this.checkAndModifyRoomName(this.state.roomName);
+    }
+
+    /**
+     * Called when the "Create Room" button is pressed.
+     */
+    onSubmit() {
+        this.setState({isLoading: true});
+        createPolicyRoom(this.state.policyID, this.state.roomName, this.state.visibility)
+            .then(() => this.setState({isLoading: false}));
```

Comment:
is there a way this could fail?

Reply:
+1, we should probably have a fail case here, right?

Reply:
Absolutely, good catch! I think this can only really fail if the user's internet connection has issues. I'll update the promise returned by `createPolicyRoom` to also call `catch` on any errors.